graphql_compiler/query_pagination/query_parameterizer.py

```diff
@@ -1,61 +1,194 @@
+                        removed_parameters.append(_get_binary_filter_node_parameter(directive))
+                    else:
+                        new_directives.append(directive)
+                new_directives.append(directive_to_add)
+                new_selection_ast.directives = new_directives
+            new_selections.append(new_selection_ast)
+        if not found_field:
+            new_selections.insert(
+                0, FieldNode(name=NameNode(value=pagination_field), directives=[directive_to_add])
+            )
+    else:
+        # Recurse until the target vertex is reached.
+        if query_ast.selection_set is None:
+            raise AssertionError()
+
+        found_field = False
+        new_selections = []
+        for selection_ast in query_ast.selection_set.selections:
+            new_selection_ast = selection_ast
+            field_name = get_ast_field_name(selection_ast)
+            if field_name == query_path[0]:
+                found_field = True
+                new_selection_ast, sub_removed_parameters = _add_pagination_filter(
+                    selection_ast, query_path[1:], pagination_field, directive_to_add
+                )
+                removed_parameters.extend(sub_removed_parameters)
+            new_selections.append(new_selection_ast)
+
+        if not found_field:
+            raise AssertionError()
+
+    new_ast = copy(query_ast)
+    new_ast.selection_set = SelectionSetNode(selections=new_selections)
+    return new_ast, removed_parameters
+
+
+def _make_binary_filter_directive_node(op_name: str, param_name: str) -> DirectiveNode:
+    """Make a binary filter directive node with the given binary op_name and parameter name."""
+    return DirectiveNode(
+        name=NameNode(value="filter"),
+        arguments=[
+            ArgumentNode(name=NameNode(value="op_name"), value=StringValueNode(value=op_name),),
+            ArgumentNode(
+                name=NameNode(value="value"),
+                value=ListValueNode(values=[StringValueNode(value="$" + param_name)]),
+            ),
+        ],
+    )
+
+
+def generate_parameterized_queries(
+    schema_info: QueryPlanningSchemaInfo,
+    query: ASTWithParameters,
+    vertex_partition: VertexPartitionPlan,
+) -> Tuple[ASTWithParameters, ASTWithParameters, str]:
+    """Generate two parameterized queries that can be used to paginate over a given query.
 
     Args:
         schema_info: QueryPlanningSchemaInfo
-        query_ast: Document, query that is being paginated.
-        parameters: dict, list of parameters for the given query.
+        query: the query to parameterize
+        vertex_partition: pagination plan
 
     Returns:
-        ParameterizedPaginationQueries namedtuple
+        # TODO use ASTWithParameters struct to simplify API
+        next_page: Ast and params for next page.
+        remainder_ast: Ast and params for remainder.
+        param_name: The parameter name used in the new filters.
```

Comment:
Currently this function is independent of what value will be inserted to define the boundary between two pages. This function outputs a `param_name`, and the caller is responsible for inserting the desired value.

This flexibility comes at a cost: The caller is expected to insert a value that makes the filter stronger than any of the filters that were removed in its place. For example, if there's an existing `x > $val, where val = 5` filter, the parameterizer will remove it and in it's place put `x > $__paged_param_0`. The caller will probably use a value greater than `5` for the `__paged_param_0` parameter, since that is in the spec of the parameter generator.

However, it would be nice to assert that all filters that were removed are replaced by stronger filters. One way to do that is to pass the desired parameter value into this function. That will add a cross-dependency on parameter generation, but it's a dependency that's very easy to mock out in tests, so we can still test the modules separately.

@obi1kenobi thoughts?

Reply:
It seems reasonable to make this code assert its correctness, especially a bug there would be quite tricky to notice and fix. Let's go for adding the dependency, especially if you think it's easy to mock out.